src/htm/ntypes/ArrayBase.cpp

```diff
@@ -576,74 +576,120 @@ void ArrayBase::fromYAML(const std::string &data) { // handles both YAML and JSO
     }
     break;
   case NTA_BasicType_Int16:
     for (size_t i = 0; i < num; i++) {
-      ((Int16 *)inbuf)[i] = vm2[i].as<Int16>();
+      ((Int16 *)inbuf)[i] = vm[i].as<Int16>();
     }
     break;
   case NTA_BasicType_UInt16:
     for (size_t i = 0; i < num; i++) {
-      ((UInt16 *)inbuf)[i] = vm2[i].as<UInt16>();
+      ((UInt16 *)inbuf)[i] = vm[i].as<UInt16>();
     }
     break;
   case NTA_BasicType_Int32:
     for (size_t i = 0; i < num; i++) {
-      ((Int32 *)inbuf)[i] = vm2[i].as<Int32>();
+      ((Int32 *)inbuf)[i] = vm[i].as<Int32>();
     }
     break;
   case NTA_BasicType_UInt32:
     for (size_t i = 0; i < num; i++) {
-      ((UInt32 *)inbuf)[i] = vm2[i].as<UInt32>();
+      ((UInt32 *)inbuf)[i] = vm[i].as<UInt32>();
     }
     break;
   case NTA_BasicType_Int64:
     for (size_t i = 0; i < num; i++) {
-      ((Int64 *)inbuf)[i] = vm2[i].as<Int64>();
+      ((Int64 *)inbuf)[i] = vm[i].as<Int64>();
     }
     break;
   case NTA_BasicType_UInt64:
     for (size_t i = 0; i < num; i++) {
-      ((UInt64 *)inbuf)[i] = vm2[i].as<UInt64>();
+      ((UInt64 *)inbuf)[i] = vm[i].as<UInt64>();
     }
     break;
   case NTA_BasicType_Real32:
     for (size_t i = 0; i < num; i++) {
-      ((Real32 *)inbuf)[i] = vm2[i].as<Real32>();
+      ((Real32 *)inbuf)[i] = vm[i].as<Real32>();
     }
     break;
   case NTA_BasicType_Real64:
     for (size_t i = 0; i < num; i++) {
-      ((Real64 *)inbuf)[i] = vm2[i].as<Real64>();
+      ((Real64 *)inbuf)[i] = vm[i].as<Real64>();
     }
     break;
   case NTA_BasicType_Bool:
     for (size_t i = 0; i < num; i++) {
-      ((bool *)inbuf)[i] = vm2[i].as<bool>();
+      ((bool *)inbuf)[i] = vm[i].as<bool>();
     }
     break;
   case NTA_BasicType_SDR: //  Expecting sparse data
   {
     SDR &sdr = getSDR();
-    SDR_sparse_t sparse;
-    for (size_t i = 0; i < num; i++) {
-      UInt x = vm2[i].as<UInt>();
-      sparse.push_back(x);
+    bool isDense = false;
+    if (num == getCount()) {
+      if (num == 1) {
+        isDense = true;
+      } else if (num == 2) { // maybe problem here
+        isDense = true;
+      } else {
+        if (vm[0] == vm[1] || vm[0] == vm[2] || vm[1] == vm[2]) {
+          isDense = true;
+        }
+      }
+    }
+
+    if (isDense) {
+      SDR_dense_t dense;
+      for (size_t i = 0; i < vm.size(); i++) {
+        bool b = vm[i].as<bool>();
+        Byte x = (b) ? 1u : 0u;
+        dense.push_back(x);
+      }
+      sdr.setDense(dense);
+    } else {
+      SDR_sparse_t sparse;
+      for (size_t i = 0; i < vm.size(); i++) {
+        UInt x = vm[i].as<UInt>();
+        sparse.push_back(x);
+      }
+      sdr.setSparse(sparse);
     }
-    sdr.setSparse(sparse);
     break;
   }
-  case NTA_BasicType_Str: 
+  case NTA_BasicType_Str:
     for (size_t i = 0; i < num; i++) {
-      ((std::string *)inbuf)[i] = vm2[i].as<std::string>();
+      ((std::string *)inbuf)[i] = vm[i].as<std::string>();
     }
     break;
   default:
```

Comment:
Also, down in line 706 we should go back to generating sparse in the toJSON( ) function.